intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java

```diff
@@ -20,67 +21,67 @@
  * Sets up the listeners for IntelliJ UI events and registers the shutdown
  * listeners.
  */
-public class IntelliJListener {
+public class IntelliJListener implements Disposable {
     
     /** The editorObservable */
     private TrackingEventManager trackingEventManager;
 
     private Project project;
 
-    /** Dummy disposable, needed for EditorFactory listener */
-    private Disposable parent;
-
     private final MessageBusConnection connection;
 
     private EditorWindowListener editorWindowListener;
 
-    private GeneralActivityListener activityListener;
+    private DebuggerListener debuggerManagerListener;
+    private BreakpointListener xBreakpointListener;
+    private DebugEventListener debuggerContextListener;
+    private DebugActionListener debuggerActionListener;
 
     /** Constructor. */
     public IntelliJListener(TrackingEventManager trackingEventManager, Project project) {
         this.trackingEventManager = trackingEventManager;
         this.project = project;
 
-        parent = new Disposable() {
-            @Override
-            public void dispose() {
-                // intentionally left empty
-            }
-        };
-
-        editorWindowListener = new EditorWindowListener(project.getName());
+        editorWindowListener = new EditorWindowListener(project, trackingEventManager);
 
         final MessageBus messageBus = ApplicationManager.getApplication().getMessageBus();
         connection = messageBus.connect();
+        Disposer.register(this, connection);
+
+        this.attachListeners();
     }
 
     /**
      * Adds IntelliJ listeners including already opened windows and
      * registers shutdown and debugger listeners.
      */
-    public void attachListeners() {
+    private void attachListeners() {
         WatchDogEventType.START_IDE.process(this);
 
         connection.subscribe(ApplicationActivationListener.TOPIC,
                 new IntelliJActivationListener());
-        activityListener = new GeneralActivityListener(project.getName());
-        EditorFactory.getInstance().addEditorFactoryListener(editorWindowListener, parent);
+        Disposer.register(this, new GeneralActivityListener(project.getName()));
```

Comment:
OK, I don't understand how this line works :-). Can you explain it to me? Thanks. Why do you have to register this listener, but not the others?

Reply:
Most of the listener APIs in IntelliJ accept a `Disposable` argument. Whenever you want to add a Listener to something, you have to supply a `Disposable`. Whenever this `Disposable` is disposed, the listener is automatically removed.

Sadly, not all listener methods follow this pattern. `addEditorFactoryListener` does, but `addAWTEventListener` as used in `GeneralActivityListener` does not.

Lastly, the messageBus (`connection`) is already registered in the constructor with `connection = messageBus.connect(); Disposer.register(this, connection);`

Reply:
Wow, this non-uniformity needs to be documented somewhere

Reply:
Yup, sadly the documentation of IntelliJ itself is lackluster. I am not aware of a style guide that they use for these kind of API designs. Might be a legacy issue as well.

Reply:
Can you integrate parts of your comment above as a comment/documentation in the source code?

Reply:
Done :+1: